### ISSUE TITLE
Allow "moved" to declare resources moving between different module packages

### ISSUE THREAD

**Issue by rgooler** [hashicorp/terraform#30153]
### Terraform Version

```
Terraform v1.1.0
on darwin_amd64
```

### Terraform Configuration Files

```terraform
moved {
  from = aws_s3_bucket.demo_logs
  to   = module.s3.aws_s3_bucket.this
}

module "s3" {
  source     = "git@github.com:demo/terraform-secure-s3.git"

  # Setup the bucket as we like
  name          = "demo-logs-"
  bucket_prefix = "demo-logs-"

  lifecycle_rule = {
    id      = "delete_after_90_days"
    enabled = true

    expiration = {
      days = 90
    }
  }


  # Setup KMS as we like
  kms_allowed_services = [
    "cloudtrail.amazonaws.com"
  ]
}
```

### Expected Behavior
I should have been able to refactor this terraform into a centralized, standardized module without enforcing arbitrary code opinions upon the end user.

### Actual Behavior
An error message "Cross-package move statement" explicitly blocked this move. 

[The documentation](https://www.terraform.io/docs/language/modules/develop/refactoring.html) doesn't mention that there are limits to this kind of refactoring. [The code comments](https://github.com/hashicorp/terraform/blob/main/internal/refactoring/move_validate.go#L46-L74) provide no insight into WHY this block exists. Nor does [the blog post announcing this feature](https://www.hashicorp.com/blog/terraform-1-1-improves-refactoring-and-the-cloud-cli-experience) provide any indication that this limitation is in place, or why.

### Steps to Reproduce
* Create and apply terraform creating this resource
* Try to refactor that statement into a module in a separate repository via the `moved{}` block
* `terraform apply` and watch for the error

### Additional Context

We're generally using terraform via atlantis, where the `moved{}` system provides a major piece of functionality and reduces the need to bypass the entirety of atlantis and its configured checks in order to refactor terraform.

**Comment by crw** [comment]
Hi @rgooler, thanks for this report. The documentation does describe this limitation. See https://www.terraform.io/docs/language/modules/develop/refactoring.html#splitting-one-module-into-multiple, towards the end of the section.

> To reduce coupling between separately-packaged modules, Terraform only allows declarations of moves between modules in the same package. In other words, Terraform would not have allowed moving into module.x above if the source address of that call had not been a local path.

Until Terraform can declare version dependencies in remote references, `moved` will only be safely possible with locally-referenced sources.

**Comment by blang** [comment]
Hi @crw,
we also encountered this problem and currently evaluate vendoring all external modules to local paths for this exact reason.
Could you please elaborate on what would block a change like adding a feature flag to disable cross boundary moved blocks.
I agree that this use basically is a misuse originally intended to be used by module maintainers, but adding an option to allow cross boundary moves would make the life of many developers refactoring code to or from shared modules a lot easier.
Thx

**Comment by apparentlymart** [comment]
Hi all,

At the time I initially labelled this issue I did write out a response but I guess I must've neglected to send it, since it's clearly not there now! I'm going to try to repeat what I wrote before as well as I can remember it.

A notable change in the design of `moved` is that it's the first feature whereby we've allowed a calling module to refer directly to an object declared in a module it calls, or even indirectly to any descendant module. If designed carelessly, that feature would potentially suddenly make every resource and module call declared in a module into a compatibility constraint, without the consent of any existing module authors.

The restriction of allowing moves only within a single module package was therefore a compromise to allow shipping an initial iteration of this feature sooner, rather than blocking until there was a fully-worked design to handle cross-package moves in a way that wouldn't risk imposing new maintenance burdens on existing module authors. For modules that coexist inside a single "package", we assume that the caller and callee can change atomically in a single module release, and that there is likely to be high-bandwidth communication channel between the team maintaining the caller and the team maintaining the callee to allow for carefully planning the refactoring together.

We do intend to expand this to cross-package refactoring in a later release, and we have some draft designs which partially solve that problem, but we are intentionally waiting to learn from experiences using the first incarnation of the feature before we build on it any further, in case feedback on the current featureset inspires changes that may make the broader problem easier to solve, or at least changes how we might approach it.

Our current early design draft involves allowing module authors to explicitly declare something that, for the sake of discussion here, I'll call a "landing pad". That's just a placeholder name awaiting further research and consideration, but it's convenient to have a specific noun to use just in this comment.

The idea of a "landing pad" is to allow effectively splitting a `moved` statement into two parts: the current `moved` block syntax would live in the caller and declare what action to take, while the called module would declare only a destination where a refactoring can "land", without any statement about what might be placed there.

I'm going to show an example of a "landing pad" using a hypothetical `landing_pad` block here, but again I want to emphasize that this is just a placeholder name and in the final design this will probably have a different name:

```hcl
# NOT VALID: This is a design sketch of a possible later
# Terraform feature.

landing_pad "example" {
  to = aws_instance.example
}

resource "aws_instance" "example" {
  # (a normal aws_instance configuration)
}
```

From the caller's perspective, they'd declare a move into the landing pad rather than into the resource directly, and thus we retain the encapsulation of exactly what resource names exist inside the module:

```hcl
moved {
  from = aws_instance.example
  to   = module.foo.landing_pad.example
}

module "foo" {
  source = "registry.example.com/apparentlymart/landing-pad-example/boop"

  # ...
}
```

This therefore allows the called module to explicitly declare this external interface for callers to use, and in particular allows Terraform to detect and report the potentially-dangerous situation where you might have mismatching versions of two modules where the target of the move is too old to accept the incoming move, and thus Terraform would end up planning to destroy the object.

The reason for using a general term (for which "landing pad" is our placeholder here) instead of something specific like "move_target" is that we're intending to also expand the set of refactoring-type operations you can declare in configuration in the future, and a "landing pad" like this would also be a reasonable target for a hypothetical config-driven "import" operation, where the root module declares that it'll take ownership of some pre-existing object but bind it to an address inside a child module that wouldn't otherwise know about the specific object being imported.

There are various details of this which we still need to figure out, in addition to the naming question. For example, this design seems to work well for moving individual resource instances but it doesn't seem to scale very well to moving entire resources or even entire module calls, because both of those involve compound objects where the caller and callee both need to agree on the names/addresses of the objects nested inside.

As noted above, we're not actively trying to answer these details right now because we want to wait and get some experience with the first iteration of the feature first. However, we do intend to look into this further in a later release and finalize a design for cross-module-package refactoring, as well as other potential use-cases such as the config-driven import I mentioned above.

In the meantime, the pre-existing `terraform state mv`-based workflows are still available for any use-cases that the initial iteration of config-driven refactoring can't meet. While of course we acknowledge that those workflows are not as smooth as the new config-driven workflow (otherwise we wouldn't have implemented the config-drive workflow!), we need to take an iterative approach whenever we're exploring entirely new functionality like this.

We'll use this issue to represent the use-case, and if we have more news on this later then we'll share it here. Thanks!

**Comment by chris922** [comment]
Thanks, @apparentlymart  for this detailed explanation!

We are maintaining several TF modules and the same team is also using them, thus it would be great to have a way to accept the risk that you clearly stated and allow TF to do this move... do you think it might be possible to have a kind of flag for moves like this as a first shot where you explicitly have to state that you know the risk of doing this? (e. g. within `moved` block, CLI argument, experimental flag, ..?)

This would already be great, cause it sounds like the final solution still needs some time and we (and probably a lot of other teams) now started to refactor our codebase since `moved` is available (really loving it!).. the `terraform state mv` was previously our solution but forced us to do refactorings only when urgently required cause it's hard to manage. Thus I would really prefer to be able to do these nasty moves with `moved`.

**Comment by mariux** [comment]
Thank you, @apparentlymart, for clarifying things. 

We are hitting the same issues on a daily base as our main use-case for `moved` will be to migrate resources into modules. Most of those modules are part of our open source modules that we control but some are also from other module creators that we can not control. 

At the moment we hard-copy the modules from remote to local repository and also need to adjust for modules using modules to be able to use the `moved` feature - which is a lot of effort.

We love (:heart:) the feature and understand that this is in an early state, but we are also willing to take any risk involved as we often migrate between different modules for our customers.

Being able to define general trusted sources (or even single modules), like mentioned by @chris922, would work fine for us too, so we can just trust all our modules under [github.com/mineiros-io](github.com/mineiros-io). 

kudos for working on this and providing a solution ;)

**Comment by benoit74** [comment]
As stated above, I really need to move resources cross-packages. While waiting for a general solution makes sense, I would really love to be able to have a short-term less generic solution like a flag in the move block stating that we take the risk to have a bad dependency.
In one of my situation, we have some resources which have been added as an addition in a third-party module (the third-party module repo has been forked) because there was a bad dependency in this third-party module which couldn't be solved without a fork (I won't go into details here, maybe it was even possible, but it has been done this way anyway, and what is done is done). Now this third-party module has been cleaned-up, and we can move the additional resources out of the third-party module fork, and hence stop to maintain this shitty fork. Since this is considered a cross-module move, we can't do it easily. 
Of course we can use to the "old" `terraform mv` workflow, but this is clearly a pain with a significant risk of error since this workflow would be manual and we have to do it 60+ times.

**Comment by korinne** [comment]
Hi all! I'm the Product Manager for Terraform Core, and I'm currently writing out the requirements for this request. If anyone would be open to chatting with me about this, I'd love to speak with you. You can book time on my calendar [here](https://calendly.com/kalpers/terraform-feedback-with-product-management
), and if you'd rather chat over email, you can shoot me an email at kalpers@hashicorp.com (hopefully I don't regret advertising my email here). Otherwise, I'll review all the comments here. Thanks again for the feedback on this issue, and excited to improve it.

**Comment by apparentlymart** [comment]
The primary design goal of the configuration-driven move feature was to _make it_ the module author's job to record the refactoring they did in code, so that Terraform can resolve that automatically rather than any refactoring always being a breaking change. Without `moved` blocks, a module author must always treat any sort of refactoring as a breaking change and so in practice module authors would rarely actually do the refactorings they want to do, in order to avoid creating workflow challenges for users of their modules.

A root module author moving things into or out of a module they didn't write was not part of the scope for the initial design of the feature. However, for Terraform v1.3.0 we did some research into what it would look like to allow using the same language features to refactor into and out of modules maintained separately from the one declaring the statement and concluded, after various conversations with those who graciously gave their time to talk with us, that the potential hazards of moving between module packages are small in comparison to the benefits of being able to do so.

For that reason, Terraform v1.3.0 will now allow unrestricted use of `moved` statements between modules regardless of how they are packaged. If you do so then we strongly recommend also using version constraints (or similar exact selections for non-registry-based module sources) for the modules you are moving into or out of, so that you can be sure that your `moved` blocks will match up with resources declared in the external module. If you leave the dependencies "floating" then it's possible that future evolution of that module could cause your `moved` blocks to no longer work as intended, because the author of a third-party shared module can't possibly anticipate what all users of the module might be doing with it.

However, the most important situation this enables is when a single maintainer (or group of collaborators) is maintaining a set of module packages _intended_ for use together. In this case it's more practical to communicate about upcoming refactors and carefully coordinate changes across multiple packages to ensure that everything gets updated together in a way that leads to a valid result. The v1.3.0 relaxation is therefore, I think, primarily of interest to maintainers of a family of modules within an organization that are maintained together, which was a constituency that the original config-based refactoring design did not address. (It was focused on the kinds of modules you find on the public registry, which can and do evolve independently of their callers.)

With all of that said, then: this feature request is now addressed in the v1.3.0 release. It's currently in its release candidate phase, so unless someone who tests the release candidate raises a significant issue that causes a change of plan the next step will be the final release. Therefore I'm going to close this issue.

Thanks for the discussion here and thanks in particular to those who were willing to meet with our team synchronously to discuss the risk vs. reward tradeoff of relaxing this rule from their perspective.
